File: fastapi-python-backend/tests/error_handling_test.py
from fastapi import FastAPI
from fastapi.testclient import TestClient

from src.app import create_app


def test_returns_404_for_unknown_routes(client: TestClient) -> None:
    response = client.get('/missing')

    assert response.status_code == 404
    assert response.json()['status'] == 404
    assert response.json()['code'] == 'NOT_FOUND'
    assert response.json()['title'] == 'Not Found'
    assert '/not-found' in response.json()['type']
    assert response.json()['requestId']


def test_returns_normalized_500_payload() -> None:
    app: FastAPI = create_app()

    @app.get('/boom')
    def boom() -> None:
        raise RuntimeError('boom')

    with TestClient(app, raise_server_exceptions=False) as local_client:
        response = local_client.get('/boom')

    assert response.status_code == 500
    assert response.json()['status'] == 500
    assert response.json()['code'] == 'INTERNAL_SERVER_ERROR'
    assert response.json()['detail'] == 'Unexpected server error.'
    assert response.json()['title'] == 'Internal Server Error'
    assert response.json()['requestId']

File: hono-ts-bun-backend/tests/error-handling.test.ts
import { describe, expect, it } from "vitest";
import { createApp } from "../src/app";
import { resetRateLimitState } from "../src/middleware/rate-limit";

describe("error handling", () => {
  beforeEach(() => {
    resetRateLimitState();
  });

  it("returns 404 for unknown routes", async () => {
    const app = createApp();

    const response = await app.request("/missing");

    expect(response.status).toBe(404);
    const body = await response.json();
    expect(body.status).toBe(404);
    expect(body.code).toBe("NOT_FOUND");
    expect(body.title).toBe("Not Found");
    expect(body.type).toContain("/not-found");
    expect(body.requestId).toBeDefined();
  });

  it("returns normalized 500 payload", async () => {
    const app = createApp();
    app.get("/boom", () => {
      throw new Error("boom");
    });

    const response = await app.request("/boom");

    expect(response.status).toBe(500);
    const body = await response.json();
    expect(body.status).toBe(500);
    expect(body.code).toBe("INTERNAL_SERVER_ERROR");
    expect(body.detail).toBe("Unexpected server error.");
    expect(body.title).toBe("Internal Server Error");
    expect(body.requestId).toBeDefined();
    expect(body.traceId).toBeDefined();
  });

  it("rate limits requests with 429", async () => {
    const app = createApp();

    let lastStatus = 200;
    for (let index = 0; index < 150; index += 1) {
      const response = await app.request("/health");
      lastStatus = response.status;
      if (lastStatus === 429) {
        break;
      }
    }

    expect(lastStatus).toBe(429);
  });
});
